Fix silent incorrectness for bmm/baddmm out_dtype overload

PaulZhang12 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #164095

Add input checks like meta functions for standard ops in `ATen/native/LinearAlgebra.cpp` for the `out_dtype` variants. Fixes silent incorrectness in https://github.com/pytorch/pytorch/issues/163816

PaulZhang12 committed:
Fix silent incorrectness for bmm/baddmm out_dtype overload

[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix silent incorrectness for bmm/baddmm out_dtype overload"




[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix silent incorrectness for bmm/baddmm out_dtype overload"




[ghstack-poisoned]

ngimel reviewed (commented):
Add a test?

Skylion007 commented on aten/src/ATen/native/cuda/Blas.cpp:
```diff
@@ -1811,6 +1811,38 @@ std::optional<c10::ScalarType> out_dtype) {
 
+static void baddbmm_bmm_out_dtype_checks(const Tensor& batch1, const Tensor& batch2, const Scalar& beta, const Scalar& alpha, const at::ScalarType out_dtype, bool is_bmm, const std::optional<Tensor>& self_baddbmm = std::nullopt) {
+  // ref ATen/native/LinearAlgebra.cpp common_checks_baddbmm_bmm
+  TORCH_CHECK(batch1.dim() == 3, "batch1 must be a 3D tensor");
```
Many of these should be TORCH_CHECK_VALUE

PaulZhang12 committed:
Update on "Fix silent incorrectness for bmm/baddmm out_dtype overload"


Add input checks like meta functions for standard ops in `ATen/native/LinearAlgebra.cpp` for the `out_dtype` variants. Fixes silent incorrectness in https://github.com/pytorch/pytorch/issues/163816




[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix silent incorrectness for bmm/baddmm out_dtype overload"


Add input checks like meta functions for standard ops in `ATen/native/LinearAlgebra.cpp` for the `out_dtype` variants. Fixes silent incorrectness in https://github.com/pytorch/pytorch/issues/163816




[ghstack-poisoned]

PaulZhang12 commented:
> Add a test?

@ngimel added tests, let me know if there are any other concerns

PaulZhang12 committed:
Update on "Fix silent incorrectness for bmm/baddmm out_dtype overload"


Add input checks like meta functions for standard ops in `ATen/native/LinearAlgebra.cpp` for the `out_dtype` variants. Fixes silent incorrectness in https://github.com/pytorch/pytorch/issues/163816




[ghstack-poisoned]

PaulZhang12 committed:
Update on "Fix silent incorrectness for bmm/baddmm out_dtype overload"


Add input checks like meta functions for standard ops in `ATen/native/LinearAlgebra.cpp` for the `out_dtype` variants. Fixes silent incorrectness in https://github.com/pytorch/pytorch/issues/163816




[ghstack-poisoned]

PaulZhang12 commented:
@pytorchbot merge

ngimel commented:
Test failure is real, perhaps checks are too strict?

PaulZhang12 committed:
Update on "Fix silent incorrectness for bmm/baddmm out_dtype overload"


Add input checks like meta functions for standard ops in `ATen/native/LinearAlgebra.cpp` for the `out_dtype` variants. Fixes silent incorrectness in https://github.com/pytorch/pytorch/issues/163816




[ghstack-poisoned]

PaulZhang12 commented:
@pytorchbot merge
